stream_alert/shared/alert.py

```diff
@@ -0,0 +1,344 @@
+
+        Args:
+            common (dict): Common features from _compute_common
+            record (dict): Individual record
+
+        Returns:
+            dict: The record subset whose values differ from the common ones.
+
+        Example:
+            _compute_diffs(
+                {'abc': 123, 'nested': {'A': 1}},  # common
+                {'abc': 123, 'nested': {'A': 1, 'B': 2}  # record
+            )
+            will return {'nested': {'B': 2}}
+        """
+        if not common:
+            return record
+
+        diff = {}
+        for key, val in record.iteritems():
+            if key not in common:
+                # This key is in the record but not in the common set.
+                # Everything in the value is unique to this record.
+                diff[key] = val
+                continue
+
+            if val == common[key]:
+                # The record value is identical to the common value - no diffs here.
+                continue
+
+            if isinstance(val, dict) and isinstance(common[key], dict):
+                # The value is a dict which is not entirely in common, but maybe partially so.
+                inner_diff = cls._compute_diff(common[key], val)
+                if inner_diff:
+                    diff[key] = inner_diff
+            else:
+                # No recursion necessary - this value is definitely not in common.
+                diff[key] = val
+
+        return diff
+
+    @classmethod
+    def merge(cls, alerts):
+        """Combine a list of alerts into a new merged alert.
+
+        The caller is responsible for determining *which* alerts should be merged, this just
+        implements the merge algorithm.
+
+        Args:
+            alerts (list): List of alerts to merge.
+                These should all have the same values for their merge keys.
+
+        Returns:
+            Alert: A new alert whose record is formed by merging the records of all the alerts.
+                The merged alert outputs are a union of all outputs in the original alerts.
+                Other information (rule name, description, etc) is copied from the first alert.
+        """
+        alerts = sorted(alerts)  # Put alerts in chronological order.
+        common = cls._compute_common([alert.record for alert in alerts])
+        merged_by = {key: common[key] for key in alerts[0].merge_by_keys}
+
+        # Keys are named such that more important information is at the beginning alphabetically.
+        new_record = {
+            'AlertCount': len(alerts),
+            'AlertTimeFirst': min(alert.created for alert in alerts),
+            'AlertTimeLast': max(alert.created for alert in alerts),
+            'MergedBy': merged_by,
+            'OtherCommonKeys': {k: v for k, v in common.iteritems() if k not in merged_by},
+            'ValueDiffs': [cls._compute_diff(common, alert.record) for alert in alerts]
+        }
+
+        # Union all of the outputs together.
+        all_outputs = set()
```

Comment:
hmm I'm not sure if this was covered in the design, but I don't know how I feel about this part. maybe you can offer some clarification, but it _appears_ that similar alerts that have an output of say "whatever-bucket" can be merged with other alerts that do _not_ have the output of "whatever-bucket" and instead have an output of "foo-firehose". is this correct? If so, I don't think this should happen. I believe the merge should only occur for alerts going to the *_same exact_* outputs. maybe I'm missing it or this logic exists elsewhere, but this sorta goes against the idea of segmentation.

for instance, if we configure a specific rule that alerts on sensitive info to send to one output (that is specific for this rule's use case), I'm thinking this logic could result in that alerts landing in other outputs as well. let me know if I'm not seeing this correctly...

EDIT: after speaking with jack and continuing my review I think maybe this merging only happens per-rule and not across rules (please confirm). in which case, when would the outputs differ and need to be union-ed here?

Reply:
Correct - merging will only happen on the same rule name.

The possibility of different outputs comes from large merge windows. A rule may have a merge window of 24-72 hours, and the outputs listed in that rule (or any parameters, really) could change in that timeframe. The most conservative approach here is to send to every output which was defined at any point during the merge window, but alternatively we could just use the set of outputs in the most recent version of the rule

Reply:
I like the idea of using the _latest_ set of outputs, but could go either way on this. let's leave it as is and maybe discuss if it become a problem for some reason

Reply:
I think latest might be more clear. Otherwise, someone might remove an output and wonder why, 12 hours a later, a merged alert raised a bunch of errors for trying to deliver to an output which doesn't exist.